include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +94,883 @@ struct Basic
-  static Anisotropic<RayDirInfo> create(
-    const Isotropic<RayDirInfo> isotropic,
-    const float3 normalizedT,
-    const float normalizedB
-  )
-  {
-    Anisotropic<RayDirInfo> retval;
-    retval::Isotropic<RayDirInfo> = isotropic;
-    retval.T = normalizedT;
-    retval.B = normalizedB;
-    
-    const float3 V = retval.getDirection();
-    retval.TdotV = dot(V,retval.T);
-    retval.BdotV = dot(V,retval.B);
-
-    return retval;
-  }
-  static Anisotropic<RayDirInfo> create(const Isotropic<RayDirInfo> isotropic, const float3 normalizedT)
-  {
-    return create(isotropic,normalizedT,cross(isotropic.N,normalizedT));
-  }
-  static Anisotropic<RayDirInfo> create(const Isotropic<RayDirInfo> isotropic)
-  {
-    float2x3 TB = nbl::hlsl::frisvad(isotropic.N);
-    return create(isotropic,TB[0],TB[1]);
-  }
-
-  float3 getTangentSpaceV() {return float3(Tdot,BdotV,Isotropic<RayDirInfo>::NdotV);}
-  // WARNING: its the transpose of the old GLSL function return value!
-  float3x3 getTangentFrame() {return float3x3(T,B,Isotropic<RayDirInfo>::N);}
-
-  float3 T;
-  float3 B;
-  float3 TdotV;
-  float3 BdotV;
+    using vector_t = vector<U, 3>;
+    using matrix_t = matrix<U, 3, 3>;
+
+    // WARNING: Changed since GLSL, now arguments need to be normalized!
+    static SAnisotropic<RayDirInfo, U> create(
+        NBL_CONST_REF_ARG(SIsotropic<RayDirInfo, U>) isotropic,
+        NBL_CONST_REF_ARG(vector_t) normalizedT,
+        const U normalizedB
+    )
+    {
+        SAnisotropic<RayDirInfo, U> retval;
+        //(SIsotropic<RayDirInfo, U>) retval = isotropic;
+        retval.T = normalizedT;
+        retval.B = normalizedB;
+        
+        const vector_t V = retval.getDirection();
+        retval.TdotV = dot(V, retval.T);
+        retval.BdotV = dot(V, retval.B);
+
+        return retval;
+    }
+    static SAnisotropic<RayDirInfo, U> create(NBL_CONST_REF_ARG(SIsotropic<RayDirInfo, U>) isotropic, NBL_CONST_REF_ARG(vector_t) normalizedT)
+    {
+        return create(isotropic, normalizedT, cross(isotropic.N, normalizedT));
+    }
+    static SAnisotropic<RayDirInfo, U> create(NBL_CONST_REF_ARG(SIsotropic<RayDirInfo, U>) isotropic)
+    {
+        matrix<U, 2, 3> TB = math::frisvad<U>(isotropic.N);
+        return create(isotropic, TB[0], TB[1]);
+    }
+
+    vector_t getTangentSpaceV() { return vector_t(TdotV, BdotV, SIsotropic<RayDirInfo, U>::NdotV); }
+    // WARNING: its the transpose of the old GLSL function return value!
+    matrix_t getTangentFrame() { return matrix_t(T, B, SIsotropic<RayDirInfo, U>::N); }
+
+    vector_t T;
+    vector_t B;
+    U TdotV;
+    U BdotV;
 };
 
 }
 
 
-template<class RayDirInfo>
-struct LightSample
+#define NBL_CONCEPT_NAME Sample
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)(typename)(typename)(typename)(typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)(I)(A)(B)(U)      // I type Isotropic, A type Aniso, B type Basic (getting clunky)
```

Comment:
I'll review this later cause its pretty meaty and you need the `_type` aliases first so there's less template params.